pkg/kubelet/kubelet.go

```diff
@@ -2845,6 +2856,76 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
 	}, 100*time.Millisecond, stopCh)
 }
 
+// handleMaxContainerCheckpoints goes through the list of checkpoints of a
+// container in a pod and deletes old checkpoints if there are more than
+// 'kl.maxContainerCheckpoints'.
+func (kl *Kubelet) handleMaxContainerCheckpoints(podUID types.UID, podFullName, containerName string, options *runtimeapi.CheckpointContainerRequest) error {
+
+	if kl.maxContainerCheckpoints <= 0 {
+		// This cannot really happen as the kubelet config parser will
+		// error out early if this is <= 0. This additional error
+		// handling is mainly here to make sure tests are working.
+		return errors.New("maxContainerCheckpoints must be greater than 0")
+	}
+
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.maxContainerCheckpointsLock.Lock()
+	defer kl.maxContainerCheckpointsLock.Unlock()
+
+	checkpointDirectoryPath := kl.getCheckpointsDir()
+
+	checkpointDirectory, err := os.ReadDir(checkpointDirectoryPath)
+	if err != nil {
+		return fmt.Errorf("error reading checkpoint archive directory %s: %w", checkpointDirectoryPath, err)
+	}
+
+	regexString := fmt.Sprintf(`checkpoint-%s-%s-\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.*tar`, podFullName, containerName)
```

Comment:
can combining podFullName or containerName cause collisions? e.g. `pod=foo,containerName=bar-baz` collides with `pod=foo-bar,containerName=baz`?

can any of the variables contain special characters that need escaping in a regex?

what is the blob being matched by `.*` before `tar`?

are `:` characters problematic to use in filenames if this code will run on Windows? (this isn't clearly limited to linux-only)

would it be a good idea to partition checkpoints for pods (e.g. into subdirectories by pod uid) so we have to scan fewer files when pruning checkpoints for a container?

Reply:
> would it be a good idea to partition checkpoints for pods (e.g. into subdirectories by pod uid) so we have to scan fewer files when pruning checkpoints for a container?

If we use per-pod subdirectories, it might be good to also take into account [namespaces](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/).
For example:
```bash
/var/lib/kubelet/checkpoints/<namespace>/<pod>/<container>/<timestamp>.tar
```

Reply:
> can combining podFullName or containerName cause collisions? e.g. `pod=foo,containerName=bar-baz` collides with `pod=foo-bar,containerName=baz`?

podFullName is `<pod>_<namespace>`, but as namespace can also contain `-` this might lead to collisions.

namespace: ns
pod: pd
container: ct-ct

would be `pd_ns-ct-ct`

namespace: ns-ct
pod: pd
container: ct

would be `pd_ns-ct-ct`. Which is the same :frowning_face:  so this does not work

> can any of the variables contain special characters that need escaping in a regex?

No, per the definition of what can be in a namespace, pod, container name. I don't think so.

> what is the blob being matched by `.*` before `tar`?

time zone as created by the RFC3339

> are `:` characters problematic to use in filenames if this code will run on Windows? (this isn't clearly limited to linux-only)

I do not know.

> would it be a good idea to partition checkpoints for pods (e.g. into subdirectories by pod uid) so we have to scan fewer files when pruning checkpoints for a container?

Works for me. Is pod uid preferred or something like @rst0git suggested?

Reply:
Reworked this PR to create checkpoints in `/var/lib/kubelet/checkpoints/POD-ID/` to avoid ambiguity with the combination of namespace name, pod name and container name as discussed above.

As this is an Alpha Feature it should be acceptable to change the location of the file. Creating the checkpoint returns the location of the checkpoint and the new value is now returned to the user.

I also removed any ':' from the time stamp encode into the file name. Instead of a RFC3339 format time stamp in the file name it now uses `UnixNano()`. This way there should be no problems on Windows file systems.

Getting the time stamp of the checkpoint archive can be now done with the the help of checkpointctl:
```
$ checkpointctl show /var/lib/kubelet/checkpoints/99feecca-9852-4f81-88e6-7349a5e6ded0/checkpoint-counters_default-counter-1684479450422081548-0000000002.tar

Displaying container checkpoint data from /var/lib/kubelet/checkpoints/99feecca-9852-4f81-88e6-7349a5e6ded0/checkpoint-counters_default-counter-1684479450422081548-0000000002.tar

+-----------+--------------------------------+--------------+---------+--------------------------------+--------+------------+------------+-------------------+
| CONTAINER |             IMAGE              |      ID      | RUNTIME |            CREATED             | ENGINE |     IP     | CHKPT SIZE | ROOT FS DIFF SIZE |
+-----------+--------------------------------+--------------+---------+--------------------------------+--------+------------+------------+-------------------+
| counter   | quay.io/adrianreber/counter:38 | 50107ee5881d | runc    | 2023-05-19T06:56:46.706793828Z | CRI-O  | 10.88.0.10 | 9.0 MiB    | 2.0 KiB           |
+-----------+--------------------------------+--------------+---------+--------------------------------+--------+------------+------------+-------------------+
```

I also included a counter in the file name as suggested. The counter starts at 0 every time the kubelet starts, but in combination with `UnixNano()` the checkpoint archive file names should now be unique and sortable by name.